test/AutoDiff/sil_differentiability_witness_parse.sil

```diff
@@ -0,0 +1,73 @@
+// RUN: %target-sil-opt %s -module-name=sil_differentiability_witness_parse | %target-sil-opt -module-name=sil_differentiability_witness_parse | %FileCheck %s
+
+// Round-trip parsing and printing test.
+
+sil_stage raw
+
+import Builtin
+import Swift
+import SwiftShims
+
+// Test public non-generic function.
+// SIL differentiability witness:
+// - Has public linkage (implicit).
+// - Has no `where` clause.
+
+sil [ossa] @foo : $@convention(thin) (Float) -> Float {
+bb0(%0 : $Float):
+  return %0 : $Float
+}
+
+sil @AD__foo__jvp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float) {
+bb0(%0 : $Float):
```

Comment:
Even `undef` can be removed! You should be able to remove the entirety of the body (from `{` through `}`).

Reply:
Same for other two functions that currently have `undef`.

Reply:
Note: parsing/printing worked for bodiless "externally available" function declarations but not internal ones:

```
SIL verification failed: external declaration of internal SILFunction not allowed: F->isAvailableExternally()
In function:
// generic
sil hidden @generic : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0
```

Deserialization, however, did not work for bodiless "externally available" functions. More info coming.

Reply:
Differentiability witness deserialization does not work for bodiless functions, regardless of linkage (`public`, `public_external`, etc): `SILDeserializer::getFuncForReference(StringRef name)` returns null for bodiless original/JVP/VJP functions.

Perhaps this is due to the test `RUN` lines (i.e. round-tripping `sil-opt` on the output of `-emit-sib`).

---

I found that [vtable entry](https://github.com/apple/swift/blob/65ebee0c7cea75ed757e179b0a9898b0cb4ecde7/lib/Serialization/DeserializeSIL.cpp#L2857) and [witness table entry](https://github.com/apple/swift/blob/65ebee0c7cea75ed757e179b0a9898b0cb4ecde7/lib/Serialization/DeserializeSIL.cpp#L2996) deserialization check whether the output of `getFuncForReference(StringRef)` is null.

I also found a `SILDeserializer::getFuncForReference(StringRef name, SILType type)` overload, which always returns a `SILFunction *`, creating a declaration if it does not exist. This is used by deserialization for various instructions like `function_ref`. I'm not sure using this is correct, but we could use it if we also serialized original/JVP/VJP function type.

---

I'll defer investigation until it's deemed necessary.

Reply:
Not supporting bodiless functions is not quite right. We should be able to use any function declaration in differentiability witnesses.